File: haskell/src/Day07.hs
module Day07 where

import Data.List
import Data.List.Split
import Data.Maybe

type Bag = (String, [(Int, String)])

searchBag :: [Bag] -> String -> Bool
searchBag mapBag key =
  let children = maybe [] (map snd) $ lookup key mapBag
   in "shiny gold" `elem` children || any (searchBag mapBag) children

countBag :: [Bag] -> (Int, String) -> Int
countBag bags (num, bag) =
  let children = fromMaybe [] . lookup bag $ bags
   in num + num * foldl' (\acc x -> acc + countBag bags x) 0 children

part1 :: [Bag] -> Int
part1 bags = length . foldl' (\acc x -> if searchBag bags x then x : acc else acc) [] . map fst $ bags

part2 :: [Bag] -> Int
part2 bags =
  let gold_children = fromMaybe [] $ lookup "shiny gold" bags
   in foldl' (\acc x -> countBag bags x + acc) 0 gold_children

inputToMap :: String -> [Bag]
inputToMap txt = map parse . lines $ txt
  where
    parse :: String -> Bag
    parse txt =
      let [sourceBag, bags] = splitOn " bags contain " txt
          bag = filter (/= (-1, "")) . map parseBag . splitOn ", " $ bags
       in (sourceBag, bag)

    parseBag :: String -> (Int, String)
    parseBag "no other bags." = (-1, "")
    parseBag bag =
      let num = head $ splitOn " " bag
          bag_name = take 2 $ tail $ splitOn " " bag
       in (read num, unwords bag_name)

File: elixir/lib/advent_of_code/day_07.ex
defmodule AdventOfCode.Day07 do
  def search_bag(key, map) do
    children = Map.get(map, key) |> Enum.map(fn {_, x} -> x end)
    "shiny gold" in children or Enum.any?(children, &search_bag(&1, map))
  end

  def count_bag({num, bag}, map) do
    children = Map.get(map, bag)
    num + num * Enum.reduce(children, 0, fn x, acc -> count_bag(x, map) + acc end)
  end

  def part1(txt) do
    bags = txt |> input_to_map

    Enum.reduce(Map.keys(bags), [], fn x, acc ->
      if search_bag(x, bags), do: [x] ++ acc, else: acc
    end)
  end

  def part2(txt) do
    bags = txt |> input_to_map
    gold_children = Map.get(bags, "shiny gold")
    Enum.reduce(gold_children, 0, fn x, acc -> count_bag(x, bags) + acc end)
  end

  @regex ~r/((?<s_bag>\w+ \w+) \w+) contain (?<bags>((\d \w+ \w+ \w+)(, )?)*|(\w+ \w+ \w+))?/
  @bags_regex ~r/(((?<bag>\d \w+ \w+) \w+)(, )*)?/
  @split_regex ~r/(?<num>\d)/
  def input_to_map(txt) do
    txt
    |> String.split("\n")
    |> Enum.map(&Regex.named_captures(@regex, &1))
    |> Enum.map(fn %{"s_bag" => sbag, "bags" => bags} ->
      bags =
        Regex.scan(@bags_regex, bags, capture: :all_names)
        |> Enum.filter(&(&1 != [""]))

      {sbag, List.flatten(bags)}
    end)
    |> Enum.map(fn {sbag, x} ->
      {sbag,
       Enum.map(x, fn x ->
         [num, bag] = Regex.split(@split_regex, x, include_captures: true, trim: true)
         {String.to_integer(num), String.trim(bag)}
       end)}
    end)
    |> Map.new()
  end
end
